### ISSUE TITLE
slide generator plugins

### ISSUE THREAD

**Issue by blahah** [slidewinder/slidewinder#40]
When a user wants to create a new slide, we give them some options.

One can be to start from scratch.
The rest can be plugins that generate a particular kind of slide or generate slides from a particular input.

For example:
- from [a figure in the academic literature](https://github.com/slidewinder/open_access_figures)
- from a tabular data format (CSV, TSV, excel) allowing subsetting / formatting of the data
- references / citations list
- whiteboard-style, allowing the presenter to draw on the slide (freehand, or with geometric shapes like powerpoint). This could be from scratch, or starting from an existing slide of any other kind. See [this very simple example I made](http://codepen.io/Blahah/pen/zrdWvM) that demonstrates how it would work (transparent canvas overlays the base element).

**Comment by blahah** [comment]
Also could be used to produce different layouts:
- two-column with stepped progress
- title and image
- title and text
- just a list

etc.

Each one just a markdown template with some helper functions that populate it.

**Comment by TransGirlCodes** [comment]
So I had a crude template idea a while back to help user generation of slides:
Template slides are a normal md slide, but it had handlebars in it's body, associated with some of the metadata at the top. The metadata would contain default values. Metadata attribute isTemplate: true would tell slidewinder - this is a template slide. When you make another slide then, which was based on said template. You would just make a slide with no body, and with metadata which replaces the default metadata in the template, and the extra attribute useTemplate: _name of template slide_. Slidewinder could just make a copy of the template slide and splice in the metadata in place of the default values at the top - handlebars would then do the rest. I only thought about standard handlebars things - didn't think about custom helpers.

**Comment by TransGirlCodes** [comment]
A related thought, say someone exports a slide to md, and edits it. How do we track this in the DAG? If we include the parent and child data in the md file and rely on that, the user could just change the data in that file before reading it in. Is some way to watch md's handles by slidewinder needed, much like how git watches files in the directory and knows when changes have been made, and whether or not they've been committed.

**Comment by blahah** [comment]
I agree with that basic template framework using handlebars - and  I think it extends easily to having custom helpers. Perhaps a template is just a .md file and (optionally) an associated js file, contained in a directory together?

RE exports I think we just include its ID and the parent-child data. If the user edits that out, that's their fault and it loses its heritage. We can maybe put a comment or something to that effect.

**Comment by TransGirlCodes** [comment]
The template idea sounds reasonable. What I worry about with parent child id in the file is that, if they edit it, 2 things are possible: 1). They loose any heritage, the slide is a singleton/orphan or whatever. 2). Artificial heritage is created. 1) is a problem as abuse of it might lead people to take slides and claim they made them from scratch. 2). Is a problem as it could potentially screw up say, slidewinder.io's database if they tried to sync their screwed up db with it.

**Comment by blahah** [comment]
I think we will need to have some logic for resolving id conflicts, but I think the 'making a false claim over a slide' problem is one we can punt until later. In general people who create content are responsible for resolving intellectual property issues with it. As long as we make it the default behaviour to include the license and credit to the parent, it's up to users if they want to work around it.

**Comment by TransGirlCodes** [comment]
I suppose upon cloning a slide, the parent has the id of the child attached to it in the database. i.e. The child is created inside the database with a body identical to the parents. THEN is read out to md. In that case, we can probably be able to detect stuff like this, we just need to think about it.

**Comment by TransGirlCodes** [comment]
Whoops! Clicked my mouse as I stood!

**Comment by blahah** [comment]
Yup I think that's right

**Comment by TransGirlCodes** [comment]
On the thing I raised about tracking files that were exported and modified. I think I have a solution, I'll draw it up, and upload images on the Gitter or somewhere.
